controllers/managedfusionoffering_controller.go

```diff
@@ -71,21 +93,170 @@ func (r *ManagedFusionOfferingReconciler) Reconcile(ctx context.Context, req ctr
+	}
+	if err := r.reconcileSubscription(); err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := r.reconcileInstallPlan(); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
 }
 
 // This function is a placeholder for offering plugin integration
 func pluginReconcile(r *ManagedFusionOfferingReconciler) (ctrl.Result, error) {
 	return ctrl.Result{}, nil
 }
 
+func (r *ManagedFusionOfferingReconciler) reconcileOperatorGroup() error {
+	r.Log.Info("Reconciling DF Offering Operator Group")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.ocsOperatorGroup, func() error {
+		if err := r.own(r.ocsOperatorGroup); err != nil {
+			return err
+		}
+		r.ocsOperatorGroup.Spec.TargetNamespaces = []string{r.namespace}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to create/update operatorGroup: %v", err)
+	}
+	return nil
+}
+
+func (r *ManagedFusionOfferingReconciler) reconcileCatalogSource() error {
+	r.Log.Info("Reconciling DF Offering Source")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.ocsCatalogSource, func() error {
+		if err := r.own(r.ocsCatalogSource); err != nil {
+			return err
+		}
+		desired := templates.OCSCatalogSource.DeepCopy()
+		// A hook will be required to get catalog for a specific offering
+		desired.Spec.Image = "registry.redhat.io/redhat/redhat-operator-index:v4.12"
+		r.ocsCatalogSource.Spec = desired.Spec
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to create/update catalogSource: %v", err)
+	}
+	return nil
+}
+
+func (r *ManagedFusionOfferingReconciler) reconcileSubscription() error {
+	r.Log.Info("Reconciling DF Offering Subscription")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.ocsSubscription, func() error {
+		if err := r.own(r.ocsSubscription); err != nil {
+			return err
+		}
+		desired := templates.OCSSubscription.DeepCopy()
+		desired.Spec.CatalogSource = r.ocsCatalogSource.Name
+		desired.Spec.CatalogSourceNamespace = r.ocsCatalogSource.Namespace
+		// A hook will be required to get channel and startingCSV for a specific offering
+		desired.Spec.Channel = "stable-4.12"
+		desired.Spec.StartingCSV = "ocs-operator.v4.12.0"
+		r.ocsSubscription.Spec = desired.Spec
+		return nil
+
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to create/update subscription: %v", err)
+	}
+	return nil
+}
+
+func (r *ManagedFusionOfferingReconciler) reconcileInstallPlan() error {
+	r.Log.Info("Reconciling DF Operator InstallPlan")
+	ocsOperatorCSV := &opv1a1.ClusterServiceVersion{}
+	ocsOperatorCSV.Name = r.ocsSubscription.Spec.StartingCSV
+	ocsOperatorCSV.Namespace = r.namespace
+	if err := r.get(ocsOperatorCSV); err != nil {
+		if errors.IsNotFound(err) {
+			var foundInstallPlan bool
+			installPlans := &opv1a1.InstallPlanList{}
+			if err := r.list(installPlans); err != nil {
+				return err
+			}
+			for i, installPlan := range installPlans.Items {
+				if findInSlice(installPlan.Spec.ClusterServiceVersionNames, ocsOperatorCSV.Name) {
```

Comment:
Why are we searching for an install plan this way? at any given point in time, only a single install plan is "active". Also, the loop might find an old install plan which has the desired CSV in it and not the active one.

Reply:
What do you mean by a single install plan being "active"?

Reply:
This was resolved offline and we are not doing manual install plans anymore. Resolving